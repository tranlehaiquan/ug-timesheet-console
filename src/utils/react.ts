import * as React from 'react'

type IdentityFunction = <T>(fn: T) => T

export const typedMemo: IdentityFunction = React.memo

export const getDisplayName = <T extends any = any>(WrappedComponent: React.ComponentType<T>): string => {
  return WrappedComponent.displayName || WrappedComponent.name || 'Component'
}
