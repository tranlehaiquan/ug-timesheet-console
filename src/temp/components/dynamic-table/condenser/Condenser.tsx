import * as React from "react";
import { Tooltip } from "../../popups/tooltip/Tooltip";

export interface ICondenserProps {
  children?: string | React.ReactNode;
}

export const Condenser: React.FC<ICondenserProps> = ({ children }) => {
  const ref = React.useRef();
  const [overflow, setOverflow] = React.useState(false);

  const checkOverflow = () => {
    if (ref && ref.current) {
      const wrapper: HTMLElement = ref.current;
      setOverflow(wrapper.clientWidth < wrapper.scrollWidth);
    }
  };

  return (
    <Tooltip
      preventShow={!overflow}
      content={children}
      delayShow={250}
      position="top"
      className="sk-truncate"
    >
      <div
        data-sk-name="children-container"
        ref={ref}
        className="sk-truncate"
        onMouseEnter={checkOverflow}
      >
        {children}
      </div>
    </Tooltip>
  );
};
