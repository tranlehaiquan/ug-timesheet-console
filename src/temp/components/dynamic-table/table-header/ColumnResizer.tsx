import * as React from 'react'
import classnames from 'classnames'

interface IResizerProps {
  onResize: React.Dispatch<React.SetStateAction<number|string>>
  min: number
}

export const ColumnResizer: React.FC<IResizerProps> = React.memo(({ onResize, min }) => {

  const ref = React.useRef(null)
  const indicatorRef = React.useRef(null)
  const [isVisible, setIsVisible] = React.useState(false)
  const [isResizing, setIsResizing] = React.useState(false)
  const indicatorHeight = calcInitialIndicatorHeight(ref)

  const onResizeStart = (e: React.MouseEvent) => {
    e.preventDefault()
    setIsResizing(true)
    window.addEventListener('mousemove', onResizeHandler)
    window.addEventListener('mouseup', onResizeEnd)
  }

  const onResizeEnd = (e: MouseEvent) => {
    e.preventDefault()
    setIsResizing(false)
    setIsVisible(false)
    window.removeEventListener('mousemove', onResizeHandler)
    window.removeEventListener('mouseup', onResizeEnd)
  }

  const onResizeHandler = (e: MouseEvent) => {
    e.preventDefault()
    const thElement = ref.current.parentElement
    const currentWidth = thElement.offsetWidth
    const newWidth = Math.max(min, currentWidth + e.movementX)

    if (newWidth !== currentWidth) {
      const tab = ref.current.closest('table')
      tab.style.width = `${tab.offsetWidth + e.movementX}px`
      onResize(newWidth)
      indicatorRef.current.style.height = `${tab.clientHeight - 1}px`
    }
  }

  const onShowHandler = (e: React.MouseEvent) => {
    e.preventDefault()
    setIsVisible(true)
  }

  const onHideHandler = (e: React.MouseEvent) => {
    e.preventDefault()
    if (!isResizing) {
      setIsVisible(false)
    }
  }

  const visibilityClass = isVisible ? 'sk-visible' : 'sk-hidden'

  return (
    <span
      data-sk-name="sk-resize-container"
      ref={ ref }
      onMouseDown={ onResizeStart }
      onMouseEnter={ onShowHandler }
      onMouseLeave={ onHideHandler }
      className="sked-column-resizer-container"
    >
      <span
        data-sk-name="sk-resize-indicator"
        ref={ indicatorRef }
        className={ classnames('sked-column-resizer-indicator', visibilityClass) }
        style={ { height: indicatorHeight } }
      />
      <span
        data-sk-name="sk-resize-handle"
        className={ classnames('sked-column-resizer-handle', visibilityClass) }
      />
    </span>
  )
})

const calcInitialIndicatorHeight = (containerRef: React.RefObject<HTMLSpanElement>) => {
  return containerRef.current && containerRef.current.closest('table')
    ? containerRef.current.closest('table').clientHeight - 1
    : 0
}
