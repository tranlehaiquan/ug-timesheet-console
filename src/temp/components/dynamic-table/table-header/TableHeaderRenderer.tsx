import * as React from "react";
import classnames from "classnames";

import { TableConfigColumns, sortDirection } from "../DynamicTable-utils";
import { Icon } from "../../icon/Icon";
import { FormInputElement } from "../../forms/FormElements";
import { TableHead, IHeaderCell } from "../../table/Table";

interface Props<T> extends React.HTMLAttributes<HTMLElement> {
  columns: TableConfigColumns<T>[];
  sortHandler?: (
    key: string,
    customSortingFunction: (
      data: T[],
      key: string,
      order: sortDirection
    ) => T[],
    sortDirection: sortDirection
  ) => (event?: React.MouseEvent<SVGElement>) => void;
  sortBy?: string;
  sortDirection?: sortDirection;
  className?: string;
  selectAll?: boolean;
  selectable?: boolean;
  selectChecked?: boolean;
  onSelect?: (event: React.ChangeEvent<HTMLInputElement>) => void;
  resizable?: boolean;
  expandable?: boolean;
  fixedColumn?: boolean;
}

type ArrowGenericComponent = <T>(
  props: Props<T>
) => React.ReactElement<Props<T>>;

export const TableHeaderRenderer: ArrowGenericComponent = (props) => {
  const { className, selectable, expandable, fixedColumn } = props;

  return (
    <TableHead
      cells={[
        ...(selectable
          ? [
              {
                minWidth: 1,
                width: 1,
                name: "select-all",
                resizable: false,
                className: classnames({ "sked-fixed-cell-head": fixedColumn }),
                content: () => selectCell(props),
              },
            ]
          : []),
        ...dynamicCells(props),
        ...(expandable
          ? [
              {
                minWidth: 34,
                width: 34,
                name: "expand",
                resizable: false,
                className: classnames("sked-table-header-cell"),
                content: () => <></>,
              },
            ]
          : []),
      ]}
      className={className}
    />
  );
};

const selectCell: ArrowGenericComponent = ({
  selectAll,
  onSelect,
  selectChecked,
}) => (
  <span>
    {selectAll && (
      <FormInputElement
        type="checkbox"
        id="select-all"
        onChange={onSelect}
        checked={!!selectChecked}
      />
    )}
  </span>
);

const dynamicCells = <T extends {}>({
  columns,
  selectable,
  fixedColumn,
  sortBy,
  sortDirection,
  sortHandler,
}: Props<T>): IHeaderCell[] =>
  columns.map((cell: TableConfigColumns<T>, index: number) => {
    const {
      name,
      key,
      width,
      headerCellClasses,
      sortable,
      sortingFunction,
      resizable,
    } = cell;
    const label = name || (key as string);
    const isSorted = sortBy === key;
    const isFixed = fixedColumn && !selectable && index === 0;
    const isFixedWithSelect = fixedColumn && selectable && index === 0;
    return {
      resizable,
      width,
      name: label,
      className: classnames(
        "sked-table-header-cell",
        headerCellClasses || "",
        { "sked-fixed-cell-head": isFixed },
        { "sked-fixed-cell-head-selectable": isFixedWithSelect }
      ),
      content: (label) => (
        <div>
          <div
            key={`${label}-${index}`}
            className={classnames("sked-table-header-cell-label")}
          >
            {label}
            {sortable &&
              (isSorted ? (
                <div
                  data-sk-name="sort-icon"
                  className={classnames("sked-table-header-cell-icon", {
                    "sked-table-header-cell-icon--sortedBy": isSorted,
                  })}
                >
                  <Icon
                    name={sortDirection === "asc" ? "arrowUp" : "arrowDown"}
                    size={12}
                    onClick={sortHandler(
                      key as string,
                      sortingFunction,
                      sortDirection
                    )}
                  />
                </div>
              ) : (
                <div
                  data-sk-name="sort-icon"
                  className={classnames("sked-table-header-cell-icon")}
                >
                  <Icon
                    name={"arrowDown"}
                    size={12}
                    onClick={sortHandler(
                      key as string,
                      sortingFunction,
                      "desc"
                    )}
                  />
                </div>
              ))}
          </div>
          {(isFixed || isFixedWithSelect) && (
            <div className="sked-table-fixed-column-shadow-br" />
          )}
        </div>
      ),
    };
  });
