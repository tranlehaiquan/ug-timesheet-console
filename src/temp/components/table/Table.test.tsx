import * as React from "react";
import { render, shallow, mount } from "enzyme";
import { merge } from "lodash";

import {
  Table,
  TableCell,
  TableRow,
  TableBody,
  TableHead,
  TableHeaderCell,
  ITableHeaderCellProps,
  IHeaderCell,
} from "./Table";
import { Condenser } from "../dynamic-table/condenser/Condenser";
import { ColumnResizer } from "../dynamic-table/table-header/ColumnResizer";

describe("Tables", () => {
  describe("Table", () => {
    test("render", () => {
      const Component = render(
        <Table>
          <TableHead
            cells={[
              { name: "Resource", width: 300 },
              { name: "Tags" },
              { name: "Region" },
            ]}
          />
          <TableBody>
            <TableRow>
              <TableCell>Elliot Alderson</TableCell>
              <TableCell>
                Hacking, IT Security, Irony, A bit ... disrupted
              </TableCell>
              <TableCell>New York City</TableCell>
            </TableRow>
            <TableRow>
              <TableCell>Darlene Alderson</TableCell>
              <TableCell>Hacking, Connections</TableCell>
              <TableCell>New York City</TableCell>
            </TableRow>
          </TableBody>
        </Table>
      );
      expect(Component).toMatchSnapshot();
    });
  });

  describe("TableHead", () => {
    test("render", () => {
      const Component = render(
        <TableHead
          cells={[
            { name: "Resource", width: 300 },
            { name: "Tags" },
            { name: "Region" },
          ]}
        />
      );
      expect(Component).toMatchSnapshot();
    });

    test("creates correct number of cells", () => {
      const Component = mount(
        <TableHead
          cells={[
            { name: "Resource", width: 300 },
            { name: "Tags" },
            { name: "Region" },
          ]}
        />
      );
      expect(Component.find("th").length).toEqual(3);
    });

    test("applies specified width", () => {
      const Component = mount(
        <TableHead
          cells={[
            { name: "Resource Name", width: 300 },
            { name: "Tags" },
            { name: "Region" },
          ]}
        />
      );
      expect(
        Component.find('[data-sk-name="th-resource-name"]').prop("style").width
      ).toEqual(300);
    });

    test("sets min width", () => {
      const Component = mount(
        <TableHead
          cells={[
            { name: "Resource Name", width: 300 },
            { name: "Tags" },
            { name: "Region" },
          ]}
        />
      );
      expect(
        Component.find('[data-sk-name="th-tags"]').prop("style").minWidth
      ).toEqual("120px");
    });
  });

  describe("TableHeaderCell", () => {
    const sampleProps = {
      cell: {
        name: "test_column",
      },
      index: 1,
    };

    const createProps: (props?: {
      cell?: {};
      index?: number;
    }) => ITableHeaderCellProps = (props) => ({
      ...merge(sampleProps, props),
    });

    test("no resizer by default", () => {
      const component = mount(<TableHeaderCell {...createProps()} />);
      expect(component.find(ColumnResizer).length).toEqual(0);
    });

    test("creates resizer when necessary", () => {
      const props = createProps({
        cell: {
          resizable: true,
        },
      });
      const component = mount(<TableHeaderCell {...props} />);
      expect(component.find(ColumnResizer).length).toEqual(1);
    });

    test("wraps cell name in a condenser", () => {
      const component = mount(<TableHeaderCell {...createProps()} />);
      const condenser = component.find(Condenser);
      expect(condenser.length).toEqual(1);
      expect(condenser.text()).toEqual(sampleProps.cell.name);
    });

    test("keeps th's css position non-static", () => {
      const component = mount(<TableHeaderCell {...createProps()} />);
      const th = component.find("th");
      expect(th.length).toBe(1);
      expect(th.props().style.position).toEqual("sticky");
    });

    test("sets default condenser min width to 120", () => {
      const props = createProps({
        cell: {
          resizable: true,
        },
      });
      const component = mount(<TableHeaderCell {...props} />);
      const resizer = component.find(ColumnResizer);
      expect(resizer.props().min).toEqual(120);
    });

    test("applies minWidth setting to resizer", () => {
      const props = createProps({
        cell: {
          resizable: true,
          minWidth: 200,
        },
      });
      const component = mount(<TableHeaderCell {...props} />);
      const resizer = component.find(ColumnResizer);
      expect(resizer.props().min).toEqual(props.cell.minWidth);
    });
  });

  describe("TableBody", () => {
    test("render", () => {
      const Component = render(
        <TableBody>
          <TableRow>
            <TableCell>Cell content</TableCell>
          </TableRow>
        </TableBody>
      );
      expect(Component).toMatchSnapshot();
    });
  });

  describe("TableRow", () => {
    test("render", () => {
      const Component = render(
        <TableRow>
          <TableCell>Cell content</TableCell>
        </TableRow>
      );
      expect(Component).toMatchSnapshot();
    });
  });

  describe("TableCell", () => {
    test("render", () => {
      const Component = render(<TableCell>Cell content</TableCell>);
      expect(Component).toMatchSnapshot();
    });
  });
});
