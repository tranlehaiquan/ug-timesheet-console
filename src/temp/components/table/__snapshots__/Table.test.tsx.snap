// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tables Table render 1`] = `
<table
  class="sk-w-full sk-border-collapse sk-table-fixed"
>
  <thead
    class="sk-bt-2 sk-bb-2 sk-border-t sk-border-b sk-border-grey-light sk-bg-grey-lightest sk-text-navy-lighter sk-text-xxs sk-uppercase"
  >
    <tr
      class="sk-h-10 sk-text-left"
    >
      <th
        class="sk-px-5 sk-font-medium"
        data-sk-name="th-resource"
        style="width:300px;min-width:120px;position:sticky"
      >
        <div
          class="sked-infowindow__trigger sk-truncate"
          data-sk-name="sked-infowindow-trigger"
        >
          <div
            class="sk-truncate"
            data-sk-name="children-container"
          >
            Resource
          </div>
        </div>
      </th>
      <th
        class="sk-px-5 sk-font-medium"
        data-sk-name="th-tags"
        style="width:undefinedpx;min-width:120px;position:sticky"
      >
        <div
          class="sked-infowindow__trigger sk-truncate"
          data-sk-name="sked-infowindow-trigger"
        >
          <div
            class="sk-truncate"
            data-sk-name="children-container"
          >
            Tags
          </div>
        </div>
      </th>
      <th
        class="sk-px-5 sk-font-medium"
        data-sk-name="th-region"
        style="width:undefinedpx;min-width:120px;position:sticky"
      >
        <div
          class="sked-infowindow__trigger sk-truncate"
          data-sk-name="sked-infowindow-trigger"
        >
          <div
            class="sk-truncate"
            data-sk-name="children-container"
          >
            Region
          </div>
        </div>
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      class="sk-border-b sk-border-grey-lighter"
    >
      <td
        class="sk-py-3 sk-px-5 sk-text-xs"
      >
        Elliot Alderson
      </td>
      <td
        class="sk-py-3 sk-px-5 sk-text-xs"
      >
        Hacking, IT Security, Irony, A bit ... disrupted
      </td>
      <td
        class="sk-py-3 sk-px-5 sk-text-xs"
      >
        New York City
      </td>
    </tr>
    <tr
      class="sk-border-b sk-border-grey-lighter"
    >
      <td
        class="sk-py-3 sk-px-5 sk-text-xs"
      >
        Darlene Alderson
      </td>
      <td
        class="sk-py-3 sk-px-5 sk-text-xs"
      >
        Hacking, Connections
      </td>
      <td
        class="sk-py-3 sk-px-5 sk-text-xs"
      >
        New York City
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Tables TableBody render 1`] = `
<tbody>
  <tr
    class="sk-border-b sk-border-grey-lighter"
  >
    <td
      class="sk-py-3 sk-px-5 sk-text-xs"
    >
      Cell content
    </td>
  </tr>
</tbody>
`;

exports[`Tables TableCell render 1`] = `
<td
  class="sk-py-3 sk-px-5 sk-text-xs"
>
  Cell content
</td>
`;

exports[`Tables TableHead render 1`] = `
<thead
  class="sk-bt-2 sk-bb-2 sk-border-t sk-border-b sk-border-grey-light sk-bg-grey-lightest sk-text-navy-lighter sk-text-xxs sk-uppercase"
>
  <tr
    class="sk-h-10 sk-text-left"
  >
    <th
      class="sk-px-5 sk-font-medium"
      data-sk-name="th-resource"
      style="width:300px;min-width:120px;position:sticky"
    >
      <div
        class="sked-infowindow__trigger sk-truncate"
        data-sk-name="sked-infowindow-trigger"
      >
        <div
          class="sk-truncate"
          data-sk-name="children-container"
        >
          Resource
        </div>
      </div>
    </th>
    <th
      class="sk-px-5 sk-font-medium"
      data-sk-name="th-tags"
      style="width:undefinedpx;min-width:120px;position:sticky"
    >
      <div
        class="sked-infowindow__trigger sk-truncate"
        data-sk-name="sked-infowindow-trigger"
      >
        <div
          class="sk-truncate"
          data-sk-name="children-container"
        >
          Tags
        </div>
      </div>
    </th>
    <th
      class="sk-px-5 sk-font-medium"
      data-sk-name="th-region"
      style="width:undefinedpx;min-width:120px;position:sticky"
    >
      <div
        class="sked-infowindow__trigger sk-truncate"
        data-sk-name="sked-infowindow-trigger"
      >
        <div
          class="sk-truncate"
          data-sk-name="children-container"
        >
          Region
        </div>
      </div>
    </th>
  </tr>
</thead>
`;

exports[`Tables TableRow render 1`] = `
<tr
  class="sk-border-b sk-border-grey-lighter"
>
  <td
    class="sk-py-3 sk-px-5 sk-text-xs"
  >
    Cell content
  </td>
</tr>
`;
